File: mcomix/thumbnailer.py
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program. If not, see <https://www.gnu.org/licenses/>.

from pathlib import Path

from gi.repository import GdkPixbuf

from loguru import logger

import mcomix.image_tools as image_tools

class Thumbnailer:
    def __init__(self):
        """
        The Thumbnailer class is responsible for thumbnail creation.

        :param size: The dimensions for the created thumbnails (width, height).
        """

        super().__init__()

    def __call__(self, size: int, filepath: Path):
        """
        Returns a thumbnail pixbuf for <filepath>, transparently handling
        both normal image files and archives. Returns None if thumbnail creation
        failed, or if the thumbnail creation is run asynchrounosly

        :param size: The max size of any one side for the created thumbnails.
        :param filepath: Path to the image that the thumbnail is generated from.
        """

        try:
            pixbuf = GdkPixbuf.Pixbuf.new_from_file(str(filepath))
        except Exception as ex:
            logger.error(f'Failed to create thumbnail for image: \'{filepath}\'')
            logger.error(f'Exception: {ex}')
            return None

        original_width = pixbuf.get_width()
        original_height = pixbuf.get_height()

        # Calculate the new dimensions while preserving the aspect ratio
        aspect_ratio = original_width / original_height
        if original_width > original_height:
            width = size
            height = int(size / aspect_ratio)
        else:
            height = size
            width = int(size * aspect_ratio)

        pixbuf = image_tools.fit_pixbuf_to_rectangle(pixbuf, width, height, 0)
        pixbuf = image_tools.add_border_pixbuf(pixbuf)

        return pixbuf

File: src/gui/lib/thumbnailer.cxx
/**
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <https://www.gnu.org/licenses/>.
 */

#include <stop_token>

#include <glibmm.h>

#include <ztd/ztd.hxx>

#include "gui/lib/image-tools.hxx"
#include "gui/lib/thumbnailer.hxx"

void
gui::lib::thumbnailer::request(const request_data& request) noexcept
{
    {
        std::lock_guard<std::mutex> lock(this->mutex_);
        this->queue_.push(request);
    }
    this->cv_.notify_one();
}

void
gui::lib::thumbnailer::run(const std::stop_token& stoken) noexcept
{
    while (!stoken.stop_requested())
    {
        this->run_once(stoken);
    }
}

void
gui::lib::thumbnailer::run_once(const std::stop_token& stoken) noexcept
{
    request_data request;
    {
        std::unique_lock<std::mutex> lock(this->mutex_);
        this->cv_.wait(lock, stoken, [this] { return !queue_.empty(); });

        if (stoken.stop_requested())
        {
            return;
        }

        if (!this->queue_.empty())
        {
            request = this->queue_.front();
            this->queue_.pop();
        }
    }

    auto pixbuf = gui::lib::image_tools::create_thumbnail(request.file, request.thumb_size);

    if (!stoken.stop_requested())
    {
        this->signal_thumbnail_created().emit(request.page, pixbuf);
    }
}
